hcloud/instances.go

```diff
@@ -20,48 +20,99 @@ import (
 }
 
-// lookupServer attempts to locate the corresponding hcloud.Server for a given corev1.Node
+// lookupServer attempts to locate the corresponding [*hcloud.Server] or [*robotmodels.Server] for a given [*corev1.Node].
 // It returns an error if the Node has an invalid provider ID or if API requests failed.
-// It can return a nil [*hcloud.Server] if neither the ProviderID nor the Name matches an existing server.
-func (i *instances) lookupServer(ctx context.Context, node *corev1.Node) (*hcloud.Server, error) {
-	var server *hcloud.Server
+// It can return nil server if neither the ProviderID nor the Name matches an existing server.
+func (i *instances) lookupServer(
+	ctx context.Context,
+	node *corev1.Node,
+) (genericServer, error) {
 	if node.Spec.ProviderID != "" {
-		serverID, err := providerid.ToServerID(node.Spec.ProviderID)
+		var serverID int64
+		serverID, isCloudServer, err := providerid.ToServerID(node.Spec.ProviderID)
+
 		if err != nil {
 			return nil, fmt.Errorf("failed to convert provider id to server id: %w", err)
 		}
 
-		server, _, err = i.client.Server.GetByID(ctx, serverID)
+		if isCloudServer {
+			server, err := getCloudServerByID(ctx, i.client, serverID)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get hcloud server \"%d\": %w", serverID, err)
+			}
+
+			if server == nil {
+				return nil, nil
+			}
+
+			return hcloudServer{server}, nil
+		}
+
+		if i.robotClient == nil {
+			return nil, errMissingRobotCredentials
+		}
+		server, err := getRobotServerByID(i.robotClient, int(serverID), node)
 		if err != nil {
-			return nil, fmt.Errorf("failed to lookup server \"%d\": %w", serverID, err)
+			return nil, fmt.Errorf("failed to get robot server \"%d\": %w", serverID, err)
 		}
-	} else {
-		var err error
-		server, _, err = i.client.Server.GetByName(ctx, node.Name)
+
+		if server == nil {
+			return nil, nil
+		}
+
+		return robotServer{server, i.robotClient}, nil
+	}
+
+	// If the node has no provider ID we try to find the server by name from
+	// both sources. In case we find two servers, we return an error.
+	cloudServer, err := getCloudServerByName(ctx, i.client, node.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get hcloud server %q: %w", node.Name, err)
+	}
+
+	var rServer *robotmodels.Server
+	if i.robotClient != nil {
+		rServer, err = getRobotServerByName(i.robotClient, node)
 		if err != nil {
-			return nil, fmt.Errorf("failed to lookup server \"%s\": %w", node.Name, err)
+			return nil, fmt.Errorf("failed to get robot server %q: %w", node.Name, err)
 		}
 	}
-	return server, nil
+
+	switch {
+	case cloudServer != nil && rServer != nil:
+		return nil, fmt.Errorf("found both a cloud & robot server for name %q", node.Name)
+	case cloudServer != nil && rServer == nil:
+		return hcloudServer{cloudServer}, nil
+	case cloudServer == nil && rServer != nil:
+		return robotServer{rServer, i.robotClient}, nil
+	default:
+		// Both nil
+		return nil, nil
+	}
 }
```

Comment:
I dont love how this turned out. Open to any suggestions on improving it (and any sub functions)

Reply:
You could extract the first error check in a separate if, and use the switch case to only check if a var is set.